Vdc flex (#443)
terraform-providers/terraform-provider-vcd: website/docs/r/org_vdc.html.markdown

* `name` - (Required) VDC name
* `description` - (Optional) VDC friendly description
* `provider_vdc_name` - (Required, System Admin) Name of the Provider VDC from which this organization VDC is provisioned.
* `allocation_model` - (Required) The allocation model used by this VDC; must be one of 
    * AllocationVApp ("Pay as you go")
    * AllocationPool ("Allocation pool")
    * ReservationPool ("Reservation pool")

>     * Flex ("Flex") (*v2.7+*, *vCD 9.7+*)

Reviewer: I created the Flex VDC and noticed that `cpu.limit` was taken into account, but `memory.limit` was **not**:

![Screenshot 2020-02-13 at 11 37 37](https://user-images.githubusercontent.com/11330689/74421207-5399da80-4e55-11ea-80c0-c7269f9884ac.png)

Then I read in docs:

> allocated - (Optional) Capacity that is committed to be available. Value in MB or MHz. Used with **AllocationPool ("Allocation pool") and ReservationPool ("Reservation pool")**.
> limit - (Optional) Capacity limit relative to the value specified for Allocation. It must not be less than that value. If it is greater than that value, it implies over provisioning. A value of 0 specifies unlimited units. Value in MB or MHz. **Used with AllocationVApp ("Pay as you go").** 

Here we have two issues:
1. Docs need to be updated to state which and when of these should be used with the Flex model.
2. If I try to change from `memory.limit` to `memory.allocated` Terraform crashes:

```
    memory {
      // Not used with Flex
      //limit = 8000

      allocated = 8008
    }
```

Crash:

```
 # vcd_org_vdc.demo_vdc will be updated in-place
  ~ resource "vcd_org_vdc" "demo_vdc" {
        allocation_model           = "Flex"
        cpu_guaranteed             = 0.92
        cpu_speed                  = 1000
        delete_force               = true
        delete_recursive           = true
        description                = "Terraform created VDC"
        elasticity                 = true
        enable_fast_provisioning   = true
        enable_thin_provisioning   = true
        enable_vm_discovery        = false
        enabled                    = true
        id                         = "urn:vcloud:vdc:29b40a5a-93a7-496f-a02f-5c8e1e79a06b"
        include_vm_memory_overhead = false
        memory_guaranteed          = 0.91
        metadata                   = {}
        name                       = "demo-vdc"
        network_pool_name          = "vc1-TestbedCluster-06:12:07-VXLAN-NP"
        network_quota              = 5
        nic_quota                  = 0
        provider_vdc_name          = "vc1-TestbedCluster-06:12:07"
        vm_quota                   = 0

      ~ compute_capacity {
            cpu {
                allocated = 0
                limit     = 8000
                overhead  = 0
                reserved  = 0
                used      = 1000
            }

          ~ memory {
              ~ allocated = 0 -> 8008
                limit     = 0
                overhead  = 0
                reserved  = 0
                used      = 512
            }
        }

        storage_profile {
            default = true
            enabled = true
            limit   = 0
            name    = "*"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

data.vcd_vapp.demo_vapp: Refreshing state...
data.vcd_edgegateway.demo-gw: Refreshing state...
vcd_org_vdc.demo_vdc: Modifying... [id=urn:vcloud:vdc:29b40a5a-93a7-496f-a02f-5c8e1e79a06b]

Error: rpc error: code = Unavailable desc = transport is closing

Error: rpc error: code = Unavailable desc = transport is closing

Error: rpc error: code = Unavailable desc = transport is closing

panic: interface conversion: interface {} is []interface {}, not *schema.Set
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0:
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: goroutine 45 [running]:
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/terraform-providers/terraform-provider-vcd/v2/vcd.getUpdatedVdcInput(0xc00012eee0, 0xc0000de7d0, 0xc0002a6e40, 0x0, 0xc0002a6e40, 0x0)
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/terraform/vbauzysvmware/terraform-provider-vcd/vcd/resource_vcd_org_vdc.go:611 +0x17b5
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/terraform-providers/terraform-provider-vcd/v2/vcd.resourceVcdVdcUpdate(0xc00012eee0, 0x1d0d6e0, 0xc0000de7d0, 0x24, 0x2745480)
2020-02-13T11:32:35.839+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/terraform/vbauzysvmware/terraform-provider-vcd/vcd/resource_vcd_org_vdc.go:487 +0x3f4
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc000569280, 0xc000297860, 0xc00065f3e0, 0x1d0d6e0, 0xc0000de7d0, 0xc0005a9e01, 0xc00064b530, 0x1010b33)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.5.0/helper/schema/resource.go:311 +0x263
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000160b00, 0xc000171918, 0xc000297860, 0xc00065f3e0, 0xc00061f708, 0xc0000b87b8, 0x1ba48c0)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.5.0/helper/schema/provider.go:294 +0x99
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc000010080, 0x1fa0b40, 0xc000538f00, 0xc000202ba0, 0xc000010080, 0xc000538f00, 0xc00007da80)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.5.0/internal/helper/plugin/grpc_provider.go:885 +0x882
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1cde5e0, 0xc000010080, 0x1fa0b40, 0xc000538f00, 0xc000202ae0, 0x0, 0x1fa0b40, 0xc000538f00, 0xc000100400, 0x816)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.5.0/internal/tfplugin5/tfplugin5.pb.go:3189 +0x217
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000400160, 0x1fad6e0, 0xc00055f080, 0xc00015e200, 0xc0003ffaa0, 0x27197e0, 0x0, 0x0, 0x0)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:995 +0x460
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: google.golang.org/grpc.(*Server).handleStream(0xc000400160, 0x1fad6e0, 0xc00055f080, 0xc00015e200, 0x0)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1275 +0xd97
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0004fe180, 0xc000400160, 0x1fad6e0, 0xc00055f080, 0xc00015e200)
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0xbb
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: created by google.golang.org/grpc.(*Server).serveStreams.func1
2020-02-13T11:32:35.840+0200 [DEBUG] plugin.terraform-provider-vcd_v2.7.0: 	/Users/lvirbalas/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xa1
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalReadData, err: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalReadData, err: rpc error: code = Unavailable desc = transport is closing
2020-02-13T11:32:35.842+0200 [DEBUG] plugin: plugin process exited: path=/Users/lvirbalas/.terraform.d/plugins/terraform-provider-vcd_v2.7.0 pid=33071 error="exit status 2"
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [TRACE] [walkApply] Exiting eval tree: data.vcd_vapp.demo_vapp
2020/02/13 11:32:35 [TRACE] [walkApply] Exiting eval tree: data.vcd_edgegateway.demo-gw
2020/02/13 11:32:35 [DEBUG] vcd_org_vdc.demo_vdc: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [TRACE] vertex "data.vcd_edgegateway.demo-gw": visit complete
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalWriteState
2020/02/13 11:32:35 [TRACE] vertex "data.vcd_vapp.demo_vapp": visit complete
2020/02/13 11:32:35 [TRACE] EvalWriteState: writing current state object for vcd_org_vdc.demo_vdc
2020/02/13 11:32:35 [TRACE] dag/walk: upstream of "output.vapp_status" errored, so skipping
2020/02/13 11:32:35 [TRACE] dag/walk: upstream of "output.external_network" errored, so skipping
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalApplyProvisioners
2020/02/13 11:32:35 [TRACE] EvalApplyProvisioners: vcd_org_vdc.demo_vdc is not freshly-created, so no provisioning is required
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalWriteState
2020/02/13 11:32:35 [TRACE] EvalWriteState: writing current state object for vcd_org_vdc.demo_vdc
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalIf
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalIf
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalWriteDiff
2020/02/13 11:32:35 [TRACE] <root>: eval: *terraform.EvalApplyPost
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/02/13 11:32:35 [TRACE] [walkApply] Exiting eval tree: vcd_org_vdc.demo_vdc
2020/02/13 11:32:35 [TRACE] vertex "vcd_org_vdc.demo_vdc": visit complete
2020/02/13 11:32:35 [TRACE] dag/walk: upstream of "provider.vcd (close)" errored, so skipping
2020/02/13 11:32:35 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2020/02/13 11:32:35 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/02/13 11:32:35 [TRACE] statemgr.Filesystem: creating backup snapshot at terraform.tfstate.backup
2020/02/13 11:32:35 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 1165
2020/02/13 11:32:35 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2020/02/13 11:32:35 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2020/02/13 11:32:35 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
2020-02-13T11:32:35.858+0200 [DEBUG] plugin: plugin exited

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Author: fixed

* `name` - (Required) VDC name
* `description` - (Optional) VDC friendly description
* `provider_vdc_name` - (Required, System Admin) Name of the Provider VDC from which this organization VDC is provisioned.

> * `allocation_model` - (Required) The allocation model used by this VDC; must be one of {AllocationVApp ("Pay as you go"), AllocationPool ("Allocation pool"), ReservationPool ("Reservation pool"), Flex ("Flex")}

Reviewer: Let's use a system to mark options supported only in newer vCD versions here too, in the same way as we discussed here:
https://github.com/terraform-providers/terraform-provider-vcd/pull/443#discussion_r376924440

```suggestion
* `allocation_model` - (Required) The allocation model used by this VDC; must be one of {AllocationVApp ("Pay as you go"), AllocationPool ("Allocation pool"), ReservationPool ("Reservation pool"), Flex ("Flex") *v2.7+*, *vCD 9.7+*)}
```

Also, this list of models is becoming hard to read. Do the curled braces serve any purpose? Can we make it a bulleted list instead?

```suggestion
* `allocation_model` - (Required) The allocation model used by this VDC. Supported values:
    * AllocationVApp ("Pay as you go")
    * AllocationPool ("Allocation pool")
    * ReservationPool ("Reservation pool")
    * Flex ("Flex") (*v2.7+*, *vCD 9.7+*)
```

CC: @Didainius

Author: changed

Reviewer: Thanks, the resulting docs look quite nice!

![Screenshot 2020-02-12 at 12 27 38](https://user-images.githubusercontent.com/11330689/74326447-10773300-4d93-11ea-848c-603502459d0c.png)

* `network_pool_name` - (Optional, System Admin) Reference to a network pool in the Provider VDC. Required if this VDC will contain routed or isolated networks.
* `allow_over_commit` - (Optional) Set to false to disallow creation of the VDC if the AllocationModel is AllocationPool or ReservationPool and the ComputeCapacity you specified is greater than what the backing Provider VDC can supply. Default is true.
* `enable_vm_discovery` - (Optional) If true, discovery of vCenter VMs is enabled for resource pools backing this VDC. If false, discovery is disabled. If left unspecified, the actual behaviour depends on enablement at the organization level and at the system level.
* `elasticity` - (Optional, *v2.7+*) Indicates if the Flex vDC is to be elastic. Works only with Flex vCD.

> * `include_vm_memory_overhead` - (Optional, *v2.7+*) Indicate if the Flex vDC is to include memory overhead into its accounting for admission control. Works only with Flex vCD.

Reviewer: @Didainius wrote:

> One general notice. As we are making a precedent with version dependent behavior - I think we need to decide on documentation.
I would vote of having something like 9.7+ for these fields in docs - both in website, and in the schema description.

I agree. What about this form:

```suggestion
* `include_vm_memory_overhead` - (Optional, *v2.7+*, *vCD 9.7+*) Indicate if the Flex vDC is to include memory overhead into its accounting for admission control. Works only with Flex vCD.
```

Author: changed

* `network_pool_name` - (Optional, System Admin) Reference to a network pool in the Provider VDC. Required if this VDC will contain routed or isolated networks.
* `allow_over_commit` - (Optional) Set to false to disallow creation of the VDC if the AllocationModel is AllocationPool or ReservationPool and the ComputeCapacity you specified is greater than what the backing Provider VDC can supply. Default is true.
* `enable_vm_discovery` - (Optional) If true, discovery of vCenter VMs is enabled for resource pools backing this VDC. If false, discovery is disabled. If left unspecified, the actual behaviour depends on enablement at the organization level and at the system level.

> * `elasticity` - (Optional, *v2.7+*, *vCD 9.7+*) Indicates if the Flex vDC is to be elastic. Works only with Flex vCD.

Reviewer: ```suggestion
* `elasticity` - (Optional, *v2.7+*, *vCD 9.7+*) Indicates if the Flex VDC should be elastic. Works only with the Flex allocation model.
```

Author: changed

> * `include_vm_memory_overhead` - (Optional, *v2.7+*, *vCD 9.7+*) Indicate if the Flex vDC is to include memory overhead into its accounting for admission control. Works only with Flex vCD.

Reviewer: ```suggestion
* `include_vm_memory_overhead` - (Optional, *v2.7+*, *vCD 9.7+*) Indicates if the Flex VDC should include memory overhead into its accounting for admission control. Works only with the Flex allocation model.
```

Author: changed
